### PR TITLE
Add a sniff to fix duplicate full stops in comments

### PR DESCRIPTION
**Pull request by ranqiangjun**
https://www.drupal.org/project/coder/issues/3126860

### CODE REVIEW

File: coder_sniffer/Drupal/Sniffs/Commenting/DuplicateFullStopSniff.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens         = $phpcsFile->getTokens();
+        $content        = $tokens[$stackPtr]['content'];
+        $contentTrimmed = trim($content);
+
+        if (substr($contentTrimmed, -2) === '..' && substr($contentTrimmed, -3) !== '...') {
+            $phpcsFile->addFixableWarning('Duplicate full-stops found', $stackPtr, 'DuplicateFullStop');
+            if ($phpcsFile->fixer !== null) {
```
Comment (klausi) [inline coder_sniffer/Drupal/Sniffs/Commenting/DuplicateFullStopSniff.php]: please use the return value of addFixableWarning() to check if the fixer is enabled like we do in other sniffs.

File: coder_sniffer/Drupal/Sniffs/Commenting/DuplicateFullStopSniff.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                $contentReversed = strrev($content);
```
Comment (klausi) [inline coder_sniffer/Drupal/Sniffs/Commenting/DuplicateFullStopSniff.php]: why this complicated fix? Can we use strrpos() for ".." and then assemble the replacement with substr() and only add one dot?

**Review by klausi — COMMENTED** [review]
I think this sniff is not fully sufficient since it does not take comment paragraphs into account. Example:

```
// This is an example comment where I will demonstrate the usage of path ./../..
// in this sentence.
```

So I think we need to integrate this into our existing comment sniffs that check if a comment ends with a dot. DocCommentSniff, FunctionCommentSniff and InlineCommentSniff I think.

**Review by klausi — CHANGES_REQUESTED** [review]
Requesting changes per comment above.

**Comment by ranqiangjun** [comment]
Hi @klausi,  I will work on it in hours, thanks for your review!

File: tests/Drupal/Commenting/InlineCommentUnitTest.inc.fixed [hunk 3]
```diff
@@ -112,3 +112,13 @@ function test2() {
 
   return $x;
 }
+
+// An example comment where I will demonstrate the usage of path ./../..
+// in this sentence.
```
Comment (ranqiangjun) [inline tests/Drupal/Commenting/InlineCommentUnitTest.inc.fixed:117]: Addressed the example.

File: coder_sniffer/Drupal/Sniffs/Commenting/DocCommentSniff.php [hunk 4]
```diff
@@ -248,6 +248,18 @@ public function process(File $phpcsFile, $stackPtr)
             }
         }
 
+        if (strlen($shortContent) > 2 && substr($shortContent, -2) === '..') {
+            $theLast3rdChar = substr($shortContent, -3, 1);
+            // Ignore ... and /.. .
```
Comment (ranqiangjun) [inline coder_sniffer/Drupal/Sniffs/Commenting/DocCommentSniff.php:253]: Ignore comments end with three dots and comments look like a path. `/..`

**Review by ranqiangjun — COMMENTED** [review]
Self-review

File: coder_sniffer/Drupal/Sniffs/Commenting/DuplicateFullStopSniff.php (hunk 1 above)
Comment (ranqiangjun) [reply coder_sniffer/Drupal/Sniffs/Commenting/DuplicateFullStopSniff.php]: Did in this iteration.

File: coder_sniffer/Drupal/Sniffs/Commenting/DuplicateFullStopSniff.php (hunk 2 above)
Comment (ranqiangjun) [reply coder_sniffer/Drupal/Sniffs/Commenting/DuplicateFullStopSniff.php]: Here is for ignoring comments end with `...`, simplified a little in this/the new iteration

**Comment by ranqiangjun** [comment]
It's an incomplete one, will continue when I get time. Only touched DocCommentSniff and InlineCommentSniff so far.

**Comment by klausi** [comment]
Closing this pull request due to inactivity, please open a new one if you are working on this again!

**Pull request closed without merging**